ldap/client/api/src/main/java/org/apache/directory/ldap/client/api/LdapNetworkConnection.java

```diff
@@ -599,102 +604,100 @@ public boolean connect() throws LdapException
-                }
+            try
+            {
+                futures.add( connectFuture( address ) );
             }
+            catch ( InterruptedException e )
+            {
+                LOG.info( "InterruptedException occurred when connecting to " + address + " - Message: " + e.getMessage() );
+            }
+        }
 
-            // Wait for the futures to complete. Break out early if there is a successful connection.
-            breakaway:
-            while ( connFutures.size() > 0 )
+        // Wait for the futures to complete. Break out early if there is a successful connection.
+        while ( futures.size() > 0 && giveUpAfter > System.currentTimeMillis() )
+        {
+            Future<LdapNetworkConnectionFuture> completedFuture;
+            try
             {
-                for ( LdapNetworkConnectionFuture cf : connFutures )
-                {
-                    if ( !cf.getConnectFuture().isDone() )
-                    {
-                        continue;
-                    }
+                completedFuture = completionService.take();
+                futures.remove( completedFuture );
+            }
+            catch ( InterruptedException e )
+            {
+                LOG.info( "InterruptedException occurred when waiting for a future to complete" );
+                continue;
+            }
 
-                    connFutures.remove( cf );
+            LdapNetworkConnectionFuture result;
+            try
+            {
+                result = completedFuture.get();
+            }
+            catch ( InterruptedException e )
+            {
+                LOG.info( "InterruptedException occurred when fetching future results - Error: " + e.getMessage() );
+                continue;
+            }
+            catch ( ExecutionException e )
+            {
+                LOG.info( "ExecutionException occurred when fetching future results - Error: " + e.getMessage() );
+                continue;
+            }
 
-                    ConnectFuture futureResult = null;
-                    try
-                    {
-                        futureResult = cf.getConnectFuture().get();
-                    }
-                    catch ( InterruptedException e )
-                    {
-                        LOG.info( "InterruptedException occurred when fetching future results - Error: " + e.getMessage() );
-                        cf.getConnector().dispose();
-                        continue;
-                    }
-                    catch ( ExecutionException e )
-                    {
-                        LOG.info( "ExecutionException occurred when fetching future results - Error: " + e.getMessage() );
-                        cf.getConnector().dispose();
-                        continue;
-                    }
+            if ( result.getConnectFuture() == null || !result.getConnectFuture().isConnected() )
+            {
+                result.getConnector().dispose();
```

Comment:
It shouldn't according to the flow. However, it is initiated in the start as null. I will fix this.